Signal/src/ViewControllers/ImagePickerController.swift

```diff
@@ -0,0 +1,439 @@
+
+        let itemWidth = floor(availableWidth / CGFloat(itemCount))
+        let newItemSize = CGSize(width: itemWidth, height: itemWidth)
+
+        if (newItemSize != collectionViewFlowLayout.itemSize) {
+            collectionViewFlowLayout.itemSize = newItemSize
+            collectionViewFlowLayout.invalidateLayout()
+        }
+    }
+
+    // MARK: Batch Selection
+
+    lazy var doneButton: UIBarButtonItem = {
+        return UIBarButtonItem(barButtonSystemItem: .done,
+                               target: self,
+                               action: #selector(didPressDone))
+    }()
+
+    lazy var selectButton: UIBarButtonItem = {
+        return UIBarButtonItem(title: NSLocalizedString("BUTTON_SELECT", comment: "Button text to enable batch selection mode"),
+                               style: .plain,
+                               target: self,
+                               action: #selector(didTapSelect))
+    }()
+
+    var isInBatchSelectMode = false {
+        didSet {
+            collectionView!.allowsMultipleSelection = isInBatchSelectMode
+            updateSelectButton()
+            updateDoneButton()
+        }
+    }
+
+    @objc
+    func didPressDone(_ sender: Any) {
+        Logger.debug("")
+
+        guard let collectionView = self.collectionView else {
+            owsFailDebug("collectionView was unexpectedly nil")
+            return
+        }
+
+        guard let indexPaths = collectionView.indexPathsForSelectedItems else {
+            owsFailDebug("indexPaths was unexpectedly nil")
+            return
+        }
+
+        let assets: [PHAsset] = indexPaths.compactMap { return self.libraryAlbum.asset(at: $0.row) }
+        let promises = assets.map { return libraryAlbum.outgoingAttachment(for: $0) }
+        when(fulfilled: promises).map { attachments in
+            self.dismiss(animated: true) {
+                self.delegate?.imagePicker(self, didPickImageAttachments: attachments)
+            }
+        }.retainUntilComplete()
+    }
+
+    func updateDoneButton() {
+        guard let collectionView = self.collectionView else {
+            owsFailDebug("collectionView was unexpectedly nil")
+            return
+        }
+
+        if let count = collectionView.indexPathsForSelectedItems?.count, count > 0 {
+            self.doneButton.isEnabled = true
+        } else {
+            self.doneButton.isEnabled = false
+        }
+    }
+
+    func updateSelectButton() {
+        navigationItem.rightBarButtonItem = isInBatchSelectMode ? doneButton : selectButton
+    }
+
+    @objc
+    func didTapSelect(_ sender: Any) {
+        isInBatchSelectMode = true
+
+        // disabled until at least one item is selected
+        self.doneButton.isEnabled = false
+
+        // Don't allow the user to leave mid-selection, so they realized they have
+        // to cancel (lose) their selection if they leave.
+        self.navigationItem.hidesBackButton = true
```

Comment:
So if a user selects a photo, they can't hit back until they unselect that photo? That doesn't sound right, but I think I'm misunderstanding.

Reply:
Sorry, this comment is confusing because it was lifted from the MediaTileViewController. 

It's currently irrelevant anyway, because there is no back button, only a dismiss button, which isn't impeded by this behavior, but I'm going to remove it.